### PR TITLE
Use absolute path to disambiguate contract names

### LINKED ISSUE #178: Contract name clashes when the same contract name is used in multiple libraries.
A foundry project could import multiple projects such as `@openzeppelin-contracts-upgradeable` and `@openzeppelin-contracts` both contains very similar contract names (e.g., [@openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/tree/0a2cb9a445c365870ed7a8ab461b12acf3e27d63/contracts/proxy/utils/Initializable.sol) and [@openzeppelin-contracts/contracts/proxy/utils/Initializable.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/ecd2ca2cd7cac116f7a37d0e474bbb3d7d5e1c4d/contracts/proxy/utils/Initializable.sol)). Thus, the contract names would clash as we only use the simple file name as contract name internally to check whether they are the same contract.

### PR DESCRIPTION
**Pull request by nwatson22**
Closes #178 
Replaces references to contracts done by name with a name qualified by the full path of the contract, e.g. contract `Initializable` in `openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol` would get the name `openzeppelin-contracts-upgradeable%contracts%proxy%utils%Initializable`. `%` is used as a separator because the names of the contracts become directory names when the proofs are saved to disk. In generated K files the K escaping would convert this to `S2KopennzeppelinZSubcontractsZSubupgradeableZModcontractsZModproxyZModutilsZModInitializable-CONTRACT`.

When there is a duplicate name, `solc` will disambiguate the files in the `out` directory by adding their paths as subdirectories. Our `foundry.contracts` function previously would just not import these. This changes is to recursively import `.json`s directly under a `.sol` directory.

Because of this issue of the duplicates not being imported because they're in subdirectories, I don't know how to reproduce the runtime error `Project contains duplicated contract names that may clash in K definitions: {contract_name}` in master. However, I've added contracts with duplicate names under `src/tests/integration/test-data/foundry/src/` to demonstrate that duplicate test names can be imported with the new system.

### CODE REVIEW

**Comment by anvacaru** [comment]
The `quality-checks` stage is currently failing on CI. You can run `make format; make` in the repo's root to format the Python files automatically and run the unit tests.

File: src/kontrol/foundry.py [hunk 1]
```diff
@@ -283,6 +297,8 @@ def matching_sigs(self, test: str) -> list[str]:
         return test_sigs
 
     def get_test_id(self, test: str, id: int | None) -> str:
+        #          test_base_name = test.split('/')[1]
+        #          matching_proofs = self.proofs_with_test(test_base_name)
```
Comment (nwatson22) [inline src/kontrol/foundry.py:301]:
```suggestion
```

File: src/kontrol/foundry.py [hunk 2]
```diff
@@ -62,6 +62,22 @@ def __init__(
...
+            for full_contract_name in self.contracts
+            if contract_name == full_contract_name.split('%')[-1]
+        ]
+        if len(contracts) == 0:
+            raise ValueError(
+                f"Tried to look up contract name {contract_name}, found none out of {[contract_name_with_path.split('/')[-1] for contract_name_with_path in self.contracts.keys()]}"
+            )
+        if len(contracts) > 1:
+            raise ValueError(
+                f'Tried to look up duplicate contract name {contract_name}, found duplicates {[contract[0] for contract in contracts]}'
```
Comment (anvacaru) [inline src/kontrol/foundry.py:77]:
```suggestion
                f'Tried to look up contract name {contract_name}, found duplicates {[contract[0] for contract in contracts]}'
```

File: src/kontrol/foundry.py [hunk 3]
```diff
@@ -553,8 +566,12 @@ def foundry_to_dot(foundry: Foundry, test: str, version: int | None = None) -> N
 
 def foundry_list(foundry: Foundry) -> list[str]:
     all_methods = [
-        f'{contract.name}.{method.signature}' for contract in foundry.contracts.values() for method in contract.methods
+        f'{contract.name_with_path}.{method.signature}'
+        for contract in foundry.contracts.values()
+        for method in contract.methods
     ]
+    print(all_methods)
+    print(listdir(foundry.proofs_dir))
```
Comment (anvacaru) [inline src/kontrol/foundry.py:574]: Is this still required?

Comment (nwatson22) [inline src/kontrol/foundry.py:574]:
```suggestion
```

Comment (nwatson22) [reply src/kontrol/foundry.py]: Removed.

**Review by anvacaru — APPROVED** [review]

**Pull request merged**